# Домашнее задание к лекции "Базовые понятия статистики"

## Обязательная часть

Будем осуществлять работу с непростым [набором данных](https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.csv) о состоянии здоровья лошадей, испытывающих кишечные колики. 

### Задание 1. Базовое изучение

Изучить представленный набор данных на основе [описания его столбцов](https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.names) и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.

### Задание 2. Работа с выбросами

В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

### Задание 3. Работа с пропусками

Рассчитать количество выбросов для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

## Дополнительная часть (необязательная)

Выполнить задания 1-3 для всего набора данных.

In [1]:
import pandas as pd

In [2]:
horse_data = pd.read_csv('horse_data.csv', na_values = '?')
horse_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 299 entries, 0 to 298
Data columns (total 28 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   2        298 non-null    float64
 1   1        299 non-null    int64  
 2   530101   299 non-null    int64  
 3   38.50    239 non-null    float64
 4   66       275 non-null    float64
 5   28       241 non-null    float64
 6   3        243 non-null    float64
 7   3.1      230 non-null    float64
 8   ?        253 non-null    float64
 9   2.1      267 non-null    float64
 10  5        244 non-null    float64
 11  4        255 non-null    float64
 12  4.1      243 non-null    float64
 13  ?.1      196 non-null    float64
 14  ?.2      194 non-null    float64
 15  ?.3      53 non-null     float64
 16  3.2      197 non-null    float64
 17  5.1      181 non-null    float64
 18  45.00    270 non-null    float64
 19  8.40     266 non-null    float64
 20  ?.4      135 non-null    float64
 21  ?.5      102 non

In [12]:
categories = [0, 1, 3, 4, 11, 12, 17, 22]
df = horse_data.iloc[:, categories]
df.columns = ['surgery', 'age', 'rectal_temperature', 'pulse', 'peristalsis', 'abdominal_distension', 'abdomen', 'outcome']
df.head()

surgery  age  rectal_temperature  pulse  peristalsis  abdominal_distension  \
0      1.0    1                39.2   88.0          4.0                   2.0   
1      2.0    1                38.3   40.0          3.0                   1.0   
2      1.0    9                39.1  164.0          4.0                   4.0   
3      2.0    1                37.3  104.0          NaN                   NaN   
4      2.0    1                 NaN    NaN          3.0                   2.0   

   abdomen  outcome  
0      2.0      3.0  
1      1.0      1.0  
2      NaN      2.0  
3      NaN      2.0  
4      3.0      1.0

In [4]:
df.info()#проверка типов данных

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 299 entries, 0 to 298
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   surgery               298 non-null    float64
 1   age                   299 non-null    int64  
 2   rectal_temperature    239 non-null    float64
 3   pulse                 275 non-null    float64
 4   peristalsis           255 non-null    float64
 5   abdominal_distension  243 non-null    float64
 6   abdomen               181 non-null    float64
 7   outcome               298 non-null    float64
dtypes: float64(7), int64(1)
memory usage: 18.8 KB


In [4]:
df.describe() #базовое изучение

surgery        age  rectal_temperature       pulse  peristalsis  \
count  298.000000  299.00000          239.000000  275.000000   255.000000   
mean     1.395973    1.64214           38.166527   71.934545     2.913725   
std      0.489881    2.17730            0.733508   28.680522     0.976298   
min      1.000000    1.00000           35.400000   30.000000     1.000000   
25%      1.000000    1.00000           37.800000   48.000000     3.000000   
50%      1.000000    1.00000           38.200000   64.000000     3.000000   
75%      2.000000    1.00000           38.500000   88.000000     4.000000   
max      2.000000    9.00000           40.800000  184.000000     4.000000   

       abdominal_distension     abdomen     outcome  
count            243.000000  181.000000  298.000000  
mean               2.259259    3.685083    1.550336  
std                1.061472    1.492667    0.737967  
min                1.000000    1.000000    1.000000  
25%                1.000000    2.000000    1.000000  
50%                2.000000    4.000000    1.000000  
75%                3.000000    5.000000    2.000000  
max                4.000000    5.000000    3.000000

На основе этих данных, можно сделать предположение, что по возрасту лошади были молодыми (преобладание значения 1 в столбце age). В половине случаев им была проведена операция (значение 1 в столбце surgery). По количественным признакам - температура и пульс, температура была повышена (норма 37,8), также как и пульс (норма 30-40), что может быть признаком воспаления. В половине случаев животные имели меньший показатель перистальтики и легкое вздутие живота. В столбце abdomen медианное значение 4, что не соответсвует норме и говорит о хирургическом вмешательстве. Дальнейшая жизни лошади (outcome) в  половине случаев продолжалась (выжили) после болезни

2. Выбросы. В столбце age выбросом будет максимальное значение 9, т.к. предполагаются значения 1 и 2. В столбце pulse максимальное значение тоже может быть выбросом (хотя верхний порог в описании не указывается). Выбросы по возрасту можно заменить на 2, т.к. возможно это неправильная запись (в значениях только 1 и 9)

In [5]:
q1 = df.age.quantile(0.25)
q3 = df.age.quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
remove_outliers = df[df.age.between(lower_bound, upper_bound, inclusive=True)]
remove_outliers

surgery  age  rectal_temperature  pulse  peristalsis  \
0        1.0    1                39.2   88.0          4.0   
1        2.0    1                38.3   40.0          3.0   
3        2.0    1                37.3  104.0          NaN   
4        2.0    1                 NaN    NaN          3.0   
5        1.0    1                37.9   48.0          3.0   
..       ...  ...                 ...    ...          ...   
294      1.0    1                 NaN  120.0          4.0   
295      2.0    1                37.2   72.0          3.0   
296      1.0    1                37.5   72.0          4.0   
297      1.0    1                36.5  100.0          3.0   
298      1.0    1                37.2   40.0          NaN   

     abdominal_distension  abdomen  outcome  
0                     2.0      2.0      3.0  
1                     1.0      1.0      1.0  
3                     NaN      NaN      2.0  
4                     2.0      3.0      1.0  
5                     3.0      5.0      1.0  
..                    ...      ...      ...  
294                   NaN      5.0      3.0  
295                   3.0      4.0      3.0  
296                   3.0      5.0      2.0  
297                   3.0      4.0      1.0  
298                   NaN      1.0      3.0  

[275 rows x 8 columns]

In [6]:
pd.concat([df, remove_outliers]).drop_duplicates(keep=False) #кол-во выбросов по возрасту, 24 значения

surgery  age  rectal_temperature  pulse  peristalsis  \
2        1.0    9                39.1  164.0          4.0   
8        2.0    9                38.3   90.0          3.0   
12       2.0    9                38.0   92.0          3.0   
15       1.0    9                 NaN  128.0          4.0   
22       1.0    9                38.3  130.0          4.0   
38       1.0    9                39.2  146.0          NaN   
40       2.0    9                39.0  150.0          NaN   
54       1.0    9                38.6  160.0          3.0   
73       1.0    9                 NaN    NaN          NaN   
74       1.0    9                39.7  100.0          3.0   
81       1.0    9                38.1  100.0          4.0   
102      1.0    9                38.0  140.0          3.0   
108      1.0    9                38.4   84.0          3.0   
134      1.0    9                38.1  136.0          1.0   
190      2.0    9                 NaN  100.0          1.0   
211      1.0    9                38.9  120.0          3.0   
226      1.0    9                 NaN    NaN          NaN   
228      1.0    9                38.5  120.0          1.0   
243      1.0    9                38.2  124.0          3.0   
251      1.0    9                38.8  124.0          3.0   
254      1.0    9                38.8  184.0          1.0   
261      2.0    9                38.6   88.0          NaN   
274      1.0    9                38.8  150.0          3.0   
286      2.0    9                39.5   84.0          NaN   

     abdominal_distension  abdomen  outcome  
2                     4.0      NaN      2.0  
8                     1.0      NaN      1.0  
12                    2.0      1.0      2.0  
15                    3.0      5.0      2.0  
22                    NaN      NaN      1.0  
38                    NaN      NaN      2.0  
40                    NaN      NaN      1.0  
54                    4.0      NaN      2.0  
73                    NaN      NaN      2.0  
74                    NaN      NaN      3.0  
81                    1.0      NaN      1.0  
102                   2.0      1.0      1.0  
108                   3.0      NaN      2.0  
134                   3.0      NaN      2.0  
190                   1.0      NaN      1.0  
211                   3.0      NaN      1.0  
226                   NaN      NaN      1.0  
228                   NaN      NaN      1.0  
243                   4.0      NaN      1.0  
251                   4.0      4.0      2.0  
254                   3.0      NaN      2.0  
261                   NaN      NaN      1.0  
274                   2.0      NaN      2.0  
286                   NaN      NaN      1.0

In [ ]:
#заменить значения 9 на 2

In [13]:
df.age.replace('9', '2', inplace=True)

C:\Users\Admin\anaconda3\lib\site-packages\pandas\core\generic.py:6746: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._update_inplace(new_data)


In [18]:
df.describe()

surgery        age  rectal_temperature       pulse  peristalsis  \
count  298.000000  299.00000          239.000000  275.000000   255.000000   
mean     1.395973    1.64214           38.166527   71.934545     2.913725   
std      0.489881    2.17730            0.733508   28.680522     0.976298   
min      1.000000    1.00000           35.400000   30.000000     1.000000   
25%      1.000000    1.00000           37.800000   48.000000     3.000000   
50%      1.000000    1.00000           38.200000   64.000000     3.000000   
75%      2.000000    1.00000           38.500000   88.000000     4.000000   
max      2.000000    9.00000           40.800000  184.000000     4.000000   

       abdominal_distension     abdomen     outcome  
count            243.000000  181.000000  298.000000  
mean               2.259259    3.685083    1.550336  
std                1.061472    1.492667    0.737967  
min                1.000000    1.000000    1.000000  
25%                1.000000    2.000000    1.000000  
50%                2.000000    4.000000    1.000000  
75%                3.000000    5.000000    2.000000  
max                4.000000    5.000000    3.000000

In [38]:
df.dropna(thresh=7).info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 248 entries, 0 to 297
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   surgery               248 non-null    float64
 1   age                   248 non-null    int64  
 2   rectal_temperature    241 non-null    float64
 3   pulse                 244 non-null    float64
 4   peristalsis           245 non-null    float64
 5   abdominal_distension  227 non-null    float64
 6   abdomen               248 non-null    float64
 7   outcome               248 non-null    float64
dtypes: float64(7), int64(1)
memory usage: 17.4 KB


In [20]:
df.groupby(['peristalsis', 'abdominal_distension']).abdomen.median()

peristalsis  abdominal_distension
1.0          1.0                     1.0
             2.0                     3.0
             3.0                     2.0
2.0          1.0                     1.0
             2.0                     2.5
             4.0                     5.0
3.0          1.0                     2.5
             2.0                     4.0
             3.0                     5.0
             4.0                     5.0
4.0          1.0                     4.0
             2.0                     4.0
             3.0                     5.0
             4.0                     5.0
Name: abdomen, dtype: float64

In [24]:
df.abdomen.fillna(df.groupby(['peristalsis', 'abdominal_distension']).abdomen.transform('median'), inplace=True)

In [28]:
df.describe()#кол-во значений в abdomen увеличилось, mean и std не сильно изменились относительно первого прочтения

surgery        age  rectal_temperature       pulse  peristalsis  \
count  298.000000  299.00000          239.000000  275.000000   255.000000   
mean     1.395973    1.64214           38.166527   71.934545     2.913725   
std      0.489881    2.17730            0.733508   28.680522     0.976298   
min      1.000000    1.00000           35.400000   30.000000     1.000000   
25%      1.000000    1.00000           37.800000   48.000000     3.000000   
50%      1.000000    1.00000           38.200000   64.000000     3.000000   
75%      2.000000    1.00000           38.500000   88.000000     4.000000   
max      2.000000    9.00000           40.800000  184.000000     4.000000   

       abdominal_distension     abdomen     outcome  
count            243.000000  263.000000  298.000000  
mean               2.259259    3.676806    1.550336  
std                1.061472    1.390437    0.737967  
min                1.000000    1.000000    1.000000  
25%                1.000000    2.000000    1.000000  
50%                2.000000    4.000000    1.000000  
75%                3.000000    5.000000    2.000000  
max                4.000000    5.000000    3.000000

In [31]:
df.groupby(['pulse', 'peristalsis']).rectal_temperature.median()

pulse  peristalsis
36.0   2.0            38.00
       3.0            37.20
38.0   3.0            37.60
40.0   1.0            38.40
       2.0            37.95
                      ...  
140.0  4.0            35.40
150.0  3.0            38.80
160.0  3.0            38.60
164.0  4.0            39.10
184.0  1.0            38.80
Name: rectal_temperature, Length: 96, dtype: float64

In [33]:
df.rectal_temperature.fillna(df.groupby(['pulse', 'peristalsis']).rectal_temperature.transform('median'), inplace=True)

In [35]:
df.describe()

surgery        age  rectal_temperature       pulse  peristalsis  \
count  298.000000  299.00000          269.000000  275.000000   255.000000   
mean     1.395973    1.64214           38.171933   71.934545     2.913725   
std      0.489881    2.17730            0.705775   28.680522     0.976298   
min      1.000000    1.00000           35.400000   30.000000     1.000000   
25%      1.000000    1.00000           37.800000   48.000000     3.000000   
50%      1.000000    1.00000           38.150000   64.000000     3.000000   
75%      2.000000    1.00000           38.500000   88.000000     4.000000   
max      2.000000    9.00000           40.800000  184.000000     4.000000   

       abdominal_distension     abdomen     outcome  
count            243.000000  263.000000  298.000000  
mean               2.259259    3.676806    1.550336  
std                1.061472    1.390437    0.737967  
min                1.000000    1.000000    1.000000  
25%                1.000000    2.000000    1.000000  
50%                2.000000    4.000000    1.000000  
75%                3.000000    5.000000    2.000000  
max                4.000000    5.000000    3.000000

Т.к. значений всего было 299, то удалять строки с пропусками, наверно, не нужно. Поэтому можно дозаполнить через соседние значения (минимально 243 зачения в abdominal_distension)

In [42]:
df.fillna(method='bfill',inplace=True)

In [43]:
df.describe()

surgery        age  rectal_temperature       pulse  peristalsis  \
count  299.000000  299.00000          299.000000  299.000000   299.000000   
mean     1.397993    1.64214           38.168395   71.979933     2.933110   
std      0.490305    2.17730            0.722363   29.202020     0.956543   
min      1.000000    1.00000           35.400000   30.000000     1.000000   
25%      1.000000    1.00000           37.800000   48.000000     3.000000   
50%      1.000000    1.00000           38.200000   64.000000     3.000000   
75%      2.000000    1.00000           38.500000   89.000000     4.000000   
max      2.000000    9.00000           40.800000  184.000000     4.000000   

       abdominal_distension     abdomen     outcome  
count            299.000000  299.000000  299.000000  
mean               2.234114    3.705686    1.551839  
std                1.054890    1.381931    0.737187  
min                1.000000    1.000000    1.000000  
25%                1.000000    2.000000    1.000000  
50%                2.000000    4.000000    1.000000  
75%                3.000000    5.000000    2.000000  
max                4.000000    5.000000    3.000000